feat: create experiment from a funnel

joshsny opened the pull request:
## Problem

If a user creates an insight to track a metric, it's likely they want to improve that metric if they can. This supports creating an experiment from an insight. Right now we surface this option for a funnel only, and will track engagement.

## Changes

- Add a create experiment button to an insight if it can be mapped to an experiment metric directly
- Prefill the metric when creating the experiment

Loom: https://www.loom.com/share/ba32c35e3a794e08ab35eb28f8a80e56

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Tested for 3 scenarios:
- Creating from a funnels query
- Check button does not show for an invalid query
- Test an insight that fails to load
- Test an insight that has an incorrect metric for experiments
- Edit all fields that are changeable from the insight page to see that they map to the experiment

joshsny committed:
feat(wip): create experiment from insight, initial logic

joshsny commented on frontend/src/scenes/insights/InsightPageHeader.tsx:
```diff
@@ -366,6 +373,20 @@ export function InsightPageHeader({ insightLogicProps }: { insightLogicProps: In
                             </>
                         )}
+                        {insightMode !== ItemMode.Edit &&
+                            insight?.query &&
```
Note: this is not the right place for this button, should go inside the insight if possible

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

joshsny committed:
chore: move button position

joshsny committed:
Merge branch 'create-experiment-from-funnel' of https://github.com/PostHog/posthog into create-experiment-from-funnel

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

jurajmajerik commented on frontend/src/scenes/experiments/experimentLogic.tsx:
```diff
@@ -1891,3 +1916,31 @@ export function getDefaultFunnelsMetric(): ExperimentFunnelsQuery {
+    if (isFunnelsQuery(insight.query.source)) {
+        return {
+            kind: NodeKind.ExperimentFunnelsQuery,
+            funnels_query: insight.query.source,
```
Indeed, the experiment metric should not include the breakdown. The reason for this is that in the experiment view, when viewing/configuring metric, you don't want to see the breakdown yet. That's why it only gets added to the query at the results calculation time.

See:
- https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/experiments/experimentLogic.tsx#L1889 for how a Trends metric should look like
- https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/experiments/experimentLogic.tsx#L1915 for how a Funnels metric should look like

jurajmajerik replied:
However, I believe that sampling should be included, if the user is using it!

joshsny committed:
feat: create experiment from insight

joshsny committed:
chore: remove button for filters

joshsny committed:
chore: add data-attr

joshsny committed:
Merge branch 'master' into create-experiment-from-funnel

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

jurajmajerik reviewed (commented):
Cool feature, thanks for taking this on! 🙂 A few things to address:

jurajmajerik commented:
General feedback: when adding a UI feature, add screenshots and/or a Loom, makes it much easier to see how the thing should be tested.

jurajmajerik commented:
> Right now we surface this option for a funnel only, and will track engagement.

Why not add it for both right away? The [code](https://github.com/PostHog/posthog/pull/27473/files#diff-804131e5997202edfc6ab73269303be229c2c3cdc0e7e8311d0920bbbd005d92R2010) to do this is already there, so should be straightforward.

jurajmajerik commented:
Currently the `insight` value is `null` for me when navigating to the form from an insight. As a result, the desired metric doesn't get populated. The issue seems to be that somehow, the insight ID doesn't end up in the URL, from which it should be retrieved.

IMO, whenever the intention is to create an experiment from an insight, but we fail to load the insight, we should throw. Otherwise the experiment gets saved without the metric which is poor user experience.

![image](https://github.com/user-attachments/assets/2fb83db5-f08f-4cfd-a8dd-e9590e114423)

jurajmajerik commented:
![image](https://github.com/user-attachments/assets/7c5e835a-adf2-48f6-afca-bf59fa040dcf)

I personally like the position of this button, as it's prominent and will likely drive adoption for experiments. But I'd check with @PostHog/team-product-analytics if it's a good idea to put it there. Perhaps the intent of this panel is to only show properties related to the chart, so better to double check with them.

jurajmajerik commented on frontend/src/scenes/funnels/FunnelCanvasLabel.tsx:
```diff
@@ -66,6 +68,24 @@ export function FunnelCanvasLabel(): JSX.Element | null {
+                          name: (insight.name || insight.derived_name) ?? undefined,
+                      })}
+                  >
+                      Run Experiment
```
We use title case in our copy:
```suggestion
                      Run experiment
```

jurajmajerik commented on frontend/src/scenes/experiments/experimentLogic.tsx:
```diff
@@ -1034,6 +1049,12 @@ export const experimentLogic = kea<experimentLogicType>([
                 return response
             },
         },
+        insight: {
```
I suggest using a bit more specific naming for this state field, such as `queryFromInsight`, otherwise it's unclear what this refers to.

jurajmajerik commented:
> I've stripped out any properties that aren't currently available in the experiments UI

This is correct! I forgot we removed the sampling option from the metric UI a while ago :)

joshsny committed:
chore: fix type casting, running experiment for unsaved insight

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

joshsny committed:
fix: display an error toast to the user when an insight fails to load for the experiment

joshsny committed:
Merge branch 'create-experiment-from-funnel' of https://github.com/PostHog/posthog into create-experiment-from-funnel

joshsny committed:
fix: resolve type changes from merge

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

joshsny committed:
Merge branch 'master' into create-experiment-from-funnel

joshsny committed:
fix: pull in breakdown attribution value

joshsny commented:
@jurajmajerik Addressed your comments here, thanks for these. Also added a Loom showing functionality locally: [loom.com/share/ba32c35e3a794e08ab35eb28f8a80e56](https://www.loom.com/share/ba32c35e3a794e08ab35eb28f8a80e56)

zlwaterfield reviewed (commented):
Awesome, this looks great :) 

I'll fall to @jurajmajerik for approval - I like his suggestions for the tooltip and banner :)

joshsny committed:
Merge branch 'master' into create-experiment-from-funnel

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

joshsny committed:
Merge branch 'master' into create-experiment-from-funnel

joshsny committed:
chrore: rewrite metric logic to not load insights, support creating from unsaved insight

joshsny committed:
chore: add tooltip and banner

joshsny committed:
chore: use has primary metric set

joshsny committed:
chore: only let trends queries be valid

joshsny committed:
Merge branch 'master' into create-experiment-from-funnel

jurajmajerik reviewed (approved):
Nice work! 🙌

joshsny committed:
Merge branch 'master' into create-experiment-from-funnel

joshsny committed:
Merge branch 'master' into create-experiment-from-funnel

joshsny committed:
Merge branch 'master' into create-experiment-from-funnel

joshsny committed:
Merge branch 'master' into create-experiment-from-funnel

joshsny committed:
fix: move experiment metric from insight to funnel canvas to fix logic import errors

joshsny committed:
chore: move experiment metric helper functions out of logic file

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)
